source/slang/slang-ir-glsl-legalize.cpp

```diff
@@ -2732,6 +2732,49 @@ void legalizeEntryPointParameterForGLSL(
         IRInst* materialized = materializeValue(builder, globalValue);
 
         pp->replaceUsesWith(materialized);
+
+        // We finally need to replace all global variable references of a global 
+        // parameter with the actual global parameter for all function calls. 
+        // Global parameters are used with a OpStore to copy its data into a global 
+        // variable intermediary. We will follow the uses of a global parameter until 
+        // we find this OpStore, then we will replace uses of the intermediary object. 
+        IRBuilder replaceBuilder(materialized);
```

Comment:
Reading the code here, it seems the logic is:
1. Find the associating global var of `globalValue`, by searching for a Store into the var where the value is `materialized`.
2. Replace  use of globalVarToReplace with `materialized`.

I suggest getting rid of the shacky step 1 with a decoration put on the global var that references back the global parameter when we created that global var initially. Then this logic will be clean and robust.

Reply:
I think I will put the decoration on the kIROp_Param that becomes a kIROp_GlobalParam since it means we don't need some search through all GlobalInst's, same principal though.

the life time of variables to explain what I am talking about:
1. we use a `in` variable somewhere in early parsing and expression generation --> make `global variable`.

2. `GlobalVarTranslationContext` --> we make kIROp_Param & we relate kIROp_Param to a `global variable`.

3. `legalizeRayTracingEntryPointParameterForGLSL` --> we make kIROp_GlobalParam from a kRIOp_Param. We also replace uses of a `global variable` here. We know which kIROp_Param we have a connection of the variable from step `2`'.

Reply:
Yes, this is a good solution. We should be decorating IRParam to avoid violating the rules.